### PR TITLE
generate spreadsheet for allocations report

### PR DESCRIPTION
**Pull request by misaka**
### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: app/models/course.rb [hunk 1]
```diff
@@ -172,6 +174,45 @@ def allocations_report_fields
...
+  # Outputs an allocations XLSX with a randomly generated suffix to `public/`.
+  # The first argument is an optional prefix for the filename to make it
+  # easier to differentiate among ~200 other similar XLSX files. Defaults to
+  # the provider.provider_code.
+  # The filename will have a UUID prefix so that it can be hosted
+  # and linked to from an Azure bucket, but that the names can't be
+  # simply guessed.
+  # Example usage:
+  #    $ bin/rails c
+  #    > Course.save_xlsx(Course.first(20))
```
Comment (defong) [inline app/models/course.rb]:
```suggestion
  #    > Course.save_allocations_report(Course.first(20))
```

File: app/models/course.rb [hunk 2]
```diff
@@ -172,6 +174,45 @@ def allocations_report_fields
...
+  def self.save_allocations_report(file_name_prefix = all.first.provider.provider_code)
+    file_data = all.to_xlsx
+
+    file_name_suffix = SecureRandom.uuid
```
Comment (defong) [inline app/models/course.rb]: timestamp instead?

File: upload-allocations.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+echo "Getting access key"
+access_key=`az storage account keys list --account-name sadfebatallocations --subscription "DFE BAT Development" --query "[?keyName == 'key1'].value | [0]"`;
+
+for f in $files
+do
+  echo "Uploading $f"
+
+  filename=$(echo $f | sed 's/public\///g')
+
+  az storage blob upload --container-name find-allocations --name "$filename" --file "$f" --connection-string "DefaultEndpointsProtocol=https;AccountName=sadfebatallocations;AccountKey=$access_key;EndpointSuffix=core.windows.net" > /dev/null
```
Comment (defong) [inline upload-allocations.sh]: you sure?

`AccountName=sad...`

File: app/models/course.rb [hunk 4]
```diff
@@ -172,6 +174,45 @@ def allocations_report_fields
...
+  def spreadsheet_columns
+    [
+      ['Academic Year', '2020/21'],
+      ['Requested By (Name)', accrediting_provider&.provider_name],
+      ['Requested by (UKPRN)', ''],
+      ['Partner ITT Provider (Name)', provider.provider_name],
+      ['Partner ITT provider (UKPRN)', provider.organisations.first&.school_nctl_organisation&.urn],
+      ['Allocation Subject', allocation_subjects.join(' | ')],
+      ['Route', program_type],
+      ['Course Aim', qualification],
```
Comment (tvararu) [inline app/models/course.rb]: Please double check these for sanity, they are massaged from the `allocations_report_fields` above

File: app/models/course.rb [hunk 5]
```diff
@@ -172,6 +174,45 @@ def allocations_report_fields
...
+      ['Other Institution', '']
+    ]
+  end
+
+  # Outputs an allocations XLSX with a randomly generated suffix to `public/`.
+  # The first argument is an optional prefix for the filename to make it
+  # easier to differentiate among ~200 other similar XLSX files. Defaults to
+  # the provider.provider_code.
+  # The filename will have a UUID prefix so that it can be hosted
+  # and linked to from an Azure bucket, but that the names can't be
```
Comment (defong) [inline app/models/course.rb]: are you implying its a public visible bucket?

Comment (tvararu) [reply app/models/course.rb]: It's set to the `blob` setting which means yes, individual files are accessible using the link, but it's not enumerable / you can't access other files easily.

File: app/models/course.rb (hunk 1 above)
Comment (tvararu) [reply app/models/course.rb]: I will never write comments again!

File: app/models/course.rb (hunk 2 above)
Comment (tvararu) [reply app/models/course.rb]: Timestamp could be ~the same for an export of 200 or so spreadsheets, reducing its effectiveness in making the URL harder to guess 🤔

File: app/models/course.rb (hunk 5 above)
Comment (defong) [reply app/models/course.rb]: so why make it public but obfuscated?

Comment (benilovj) [reply app/models/course.rb]: we will distribute publicly available individual deep links but don't want to make it feasible to guess other file links if you don't have it.

Comment (defong) [reply app/models/course.rb]: maybe? https://blogs.msdn.microsoft.com/jpsanders/2017/10/12/easily-create-a-sas-to-download-a-file-from-azure-storage-using-azure-storage-explorer/

Comment (benilovj) [reply app/models/course.rb]: > maybe? https://blogs.msdn.microsoft.com/jpsanders/2017/10/12/easily-create-a-sas-to-download-a-file-from-azure-storage-using-azure-storage-explorer/

Yes that's definitely an option, although probably more work than above implementation, given we need to do this for 250-1000 files.

**Pull request merged**